packages/theme-check-common/src/checks/undefined-object/index.ts

```diff
@@ -0,0 +1,156 @@
+import {
+  LiquidHtmlNode,
+  LiquidTag,
+  LiquidTagAssign,
+  LiquidTagCapture,
+  LiquidTagFor,
+  LiquidTagTablerow,
+  LiquidVariableLookup,
+  NamedTags,
+  NodeTypes,
+  Position,
+} from '@shopify/liquid-html-parser';
+import { LiquidCheckDefinition, Severity, SourceCodeType } from '../../types';
+import { last } from '../../utils';
+
+type Scope = { start?: number; end?: number };
+
+export const UndefinedObject: LiquidCheckDefinition = {
+  meta: {
+    code: 'UndefinedObject',
+    name: 'Undefined Object',
+    docs: {
+      description: 'This check exists to identify references to undefined Liquid objects.',
+      recommended: true,
+      url: 'https://shopify.dev/docs/themes/tools/theme-check/checks/undefined-object',
+    },
+    type: SourceCodeType.LiquidHtml,
+    severity: Severity.WARNING,
+    schema: {},
+    targets: [],
+  },
+
+  create(context) {
+    /**
+     * At present, snippet assets are not supported due to the inability of this
+     * check to handle objects defined in other assets.
+     */
+    if (context.relativePath(context.file.absolutePath).startsWith('snippets/')) {
+      return {};
+    }
+
+    const variableScopes: Map<string, Scope[]> = new Map();
+    const variables: LiquidVariableLookup[] = [];
+
+    function indexVariableScope(variableName: string | null, scope: Scope) {
+      if (!variableName) return;
+
+      const indexedScope = variableScopes.get(variableName) ?? [];
+      const currentScope = [...indexedScope, scope];
```

Comment:
Might need a sort instead for this. Since the async nature of the check doesn't _guarantee_ that they are traversed in order.